File: squbs-unicomplex/src/test/scala/org/squbs/lifecycle/GracefulStopHelperTest.scala
/*
 *  Copyright 2017 PayPal
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.squbs.lifecycle

import akka.testkit.{TestActorRef, ImplicitSender, TestKit}
import akka.actor.{PoisonPill, ActorRef, Actor, ActorSystem}
import org.scalatest.{BeforeAndAfterAll, Matchers, FlatSpecLike}
import scala.concurrent.duration.FiniteDuration
import scala.concurrent.Future

class GracefulStopHelperTest extends TestKit(ActorSystem("testSystem")) with FlatSpecLike with Matchers with ImplicitSender with BeforeAndAfterAll {

  import system.dispatcher

  "GracefulStopHelper" should "work when stop failed" in {

    val actorRef = TestActorRef(new Actor with GracefulStopHelper {
      def receive: Actor.Receive = {
        case "Stop" =>
          defaultMidActorStop(Seq(self))
          sender() ! "Done"
      }

      override def gracefulStop(target: ActorRef, timeout: FiniteDuration, stopMessage: Any = PoisonPill): Future[Boolean] = {
        Future {
          throw new RuntimeException("BadMan")
        }

      }
    })

    val actor = actorRef.underlyingActor

    actorRef ! "Stop"
    expectMsg("Done")

  }

  "GracefulStopHelper" should "work" in {

    val actorRef = TestActorRef(new Actor with GracefulStopHelper {
      def receive: Actor.Receive = {
        case "Stop" =>
          defaultMidActorStop(Seq(self))
          sender() ! "Done"
      }

      override def gracefulStop(target: ActorRef, timeout: FiniteDuration, stopMessage: Any = PoisonPill): Future[Boolean] = {

        Future {
          true
        }

      }
    })

    val actor = actorRef.underlyingActor

    actorRef ! "Stop"
    expectMsg("Done")

  }

}

File: squbs-unicomplex/src/test/java/org/squbs/lifecycle/GracefulStopHelperTest.java
package org.squbs.lifecycle;

import org.apache.pekko.actor.ActorRef;
import org.apache.pekko.actor.ActorSystem;
import org.apache.pekko.actor.Props;
import org.apache.pekko.testkit.javadsl.TestKit;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import java.time.Duration;
import java.util.Collections;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;

public class GracefulStopHelperTest {

    static ActorSystem system;

    @BeforeClass
    public static void setUp() {
        system = ActorSystem.create("GracefulStopHelperTest");
    }

    @AfterClass
    public static void tearDown() {
        TestKit.shutdownActorSystem(system);
        system = null;
    }


    @Test
    public void testStopFailed() {
        new TestKit(system) {{
            ActorRef actor = system.actorOf(Props.create(BadShutdownActor.class));
            actor.tell("Stop", getRef());
            expectMsgEquals("Done");
        }};
    }

    @Test
    public void testStopSuccess() {
        new TestKit(system) {{
            ActorRef actor = system.actorOf(Props.create(GoodShutdownActor.class));
            actor.tell("Stop", getRef());
            expectMsgEquals("Done");
        }};
    }

    static class BadShutdownActor extends ActorWithGracefulStopHelper {

        @Override
        public Receive createReceive() {
            return receiveBuilder()
                    .matchEquals("Stop", x -> {
                        defaultMidActorStop(Collections.singletonList(getSelf()));
                        getSender().tell("Done", getSelf());
                    })
                    .build();
        }

        @Override
        public CompletionStage<Boolean> gracefulStop(ActorRef target, Duration duration, Object stopMessage) {
            return CompletableFuture.supplyAsync(() -> { throw new RuntimeException("BadMan"); });
        }
    }

    static class GoodShutdownActor extends ActorWithGracefulStopHelper {

        @Override
        public Receive createReceive() {
            return receiveBuilder()
                    .matchEquals("Stop", x -> {
                        defaultMidActorStop(Collections.singletonList(getSelf()));
                        getSender().tell("Done", getSelf());
                    })
                    .build();
        }

        @Override
        public CompletionStage<Boolean> gracefulStop(ActorRef target, Duration duration, Object stopMessage) {
            return CompletableFuture.supplyAsync(() -> true);
        }
    }
}
